### ISSUE TITLE
``used-before-assignment`` false negative with ``TYPE_CHECKING``

### ISSUE THREAD

**Issue by DanielNoord** [pylint-dev/pylint#8198]
### Bug description

Within functions ``used-before-assignment`` doesn't get raised for some edge cases.

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from datetime import datetime

# pylint: disable=missing-function-docstring, missing-module-docstring


def func_two():
    second = datetime.now()
    return second


def func():
    first: datetime
    first = datetime.now()
    second = datetime.now()
    return first, second

```

### Configuration

```ini
Default config.
```

### Command used

```shell
pylint
```

### Pylint output

```shell
************* Module test
test.py:10: [E0601(used-before-assignment), func_two] Using variable 'datetime' before assignment


-------------------------------------------------------------------
Your code has been rated at 4.55/10 (previous run: 10.00/10, -5.45)
```

### Expected behavior

```shell
************* Module test
test.py:10: [E0601(used-before-assignment), func_two] Using variable 'datetime' before assignment
test.py:16: [E0601(used-before-assignment), func_two] Using variable 'datetime' before assignment


-------------------------------------------------------------------
Your code has been rated at 4.55/10 (previous run: 10.00/10, -5.45)
```

### Pylint version

```shell
2.16.1
```

### OS / Environment

/

**Comment by jacobtylerwalls** [comment]
Is this still a problem as of #8071?

**Comment by MaximMukhametov** [comment]
> Is this still a problem as of #8071?

yes, I have the same problem with:
```python
from typing import TYPE_CHECKING
# It's needed just to avoid the PyCharm's static analyzer bug
# https://stackoverflow.com/questions/53276766/how-to-avoid-unexpected-argument-warning-using-data-classes
if TYPE_CHECKING:
    from dataclasses import dataclass  # noqa
else:
    from pydantic.dataclasses import dataclass  # noqa


@dataclass
class Item:
    """Item domain model."""

    user_id: int
    title: str
    id: int | None = None
```

**Comment by DanielNoord** [comment]
@jacobtylerwalls Yeah, on `main` this is still an issue.. I wonder if we could expand on that PR?

**Comment by zenlyj** [comment]
@DanielNoord if no one has started on this, could you assign it to me? Thanks! 🙂

**Comment by zenlyj** [comment]
@MaximMukhametov i am struggling to understand why `used-before-assignment` would be emitted in your example. Isn't `dataclass` defined by this line: `from pydantic.dataclasses import dataclass`? Please let me know if I am missing something here.

**Comment by jacobtylerwalls** [comment]
@zenlyj pylint is taking into account that `if TYPE_CHECKING:` will be false at runtime. Add the line `func()` at the bottom of Daniël's example, and execute it. You will see:
```python
    first = datetime.now()
            ^^^^^^^^
NameError: name 'datetime' is not defined
```

**Comment by zenlyj** [comment]
@jacobtylerwalls thanks for clarifying! Any thoughts on the other case posted by @MaximMukhametov? I am unsure if the `dataclass` decorator should raise `used-before-assignment`, since it is defined in the else block.

**Comment by jacobtylerwalls** [comment]
Ah, sorry! You have a point; doesn't seem like we should emit warnings in that case (and we don't currently.)
